### ISSUE TITLE
[Feature] Min/Max/Preferred Versions

### ISSUE THREAD

**Issue by mottosso** [nerdvegas/rez#631]
### Goal

Facilitate experimental and development overrides.

### Motivation

- Sometimes, asking for `rez>=2.28` and getting `2.30.0` if available is expected.
- Other times, asking for `maya` and getting `2018` despite `2019` being available is expected.

### Implementation

Normally, when you specify a minimum, it or a later available version is picked.

**pipeline/package.py**

```python
requires = ["rez>=2.28"]
```

If `2.29` is found, it gets picked. If `2.30` is found, it gets picked.

```bash
$ rez env pipeline           # rez-2.30
$ rez env pipeline rez-2.28  # rez-2.28.3
$ rez env pipeline rez-2.21  # Not allowed
```

But what about the other way?

**pipeline/package.py**

```python
requires = ["maya>=2017"]
```

In this case, I'd like for `2017` to get picked, unless a newer version is explicitly asked for.

```bash
$ rez env pipeline            # maya-2017
$ rez env pipeline maya-2019  # maya-2019
$ rez env pipeline maya-2016  # Not allowed
```

In some cases, the former is expected, and in others the latter is.

### Syntax

Some food for thought, this could be specified at the package-level.

**rez/package.py**

```python
version = "2.30.0"
order = "latest"
```

**maya/package.py**

```python
version = "2016.0.0"
order = "earliest"
```

Although for that to work we'd need to determine which of the released versions get to decide the order when two or more differ; e.g. the first found, the last released or via e.g. `rezconfig.py` which would strip away some of the nice compartmentalisation.

Furthermore, in addition to min/max, what about specifying a "preferred" version, such as..

```bash
$ rez env maya       # 2017
$ rez env maya-2016  # 2016
$ rez env maya-2018  # 2018
$ rez env maya-2015  # Not allowed
$ rez env maya-2019  # Not allowed
```

Whereby 2017 is "preferred", minimum is 2016, and maximum is 2018.

E.g.

**pipeline/package.py**

```python
requires = ["maya>=2016,<=2018"]
preferred = ["maya-2016"]
```

In this case, the order is specified in a library requiring another library which might make actually be possible without using `rezconfig.py`. A package would know whether any of its requirements are defined as min/max or preferred.

### Alternative Syntax

More food for thought, what if the request format could capture whether one wanted a latest or earliest version using plus and minus signs?

**Latest**

```bash
$ rez env maya-2018+  # 2019
```

The default, get me 2018 or later please.

**Earliest**

```bash
$ rez env maya-2018-  # 2018
```

The alternative, get me 2018 or earlier please.

**With Minimum**

```bash
$ rez env maya>2015,2018-  # 2018
```

A minimum is included, which when used as a `requires` could be used like this:

```bash
$ rez env pipeline  # maya-2018
$ rez env pipeline maya-2016  # maya-2019
$ rez env pipeline maya-2016  # maya-2016
```

**And Maximum**

```bash
$ rez env maya>2015,<2019,2018-
```

Let me know your thoughts.

**Comment by bpabel** [comment]
I don't think it makes sense to include this type of feature in the "requested" package.  It seems fairly unintuitive that another package could affect the way my version dependencies resolve.

One issue I see with the idea of using a preferred version or using a different set of version resolve rules per package dependency is what happens when these types of dependencies are encountered in the same context.

For example, I have two packages `foo` and `bar`

Using your last exampe with the min, max, preferred syntax

#### `foo`

```
requires =["maya>2015,<2019,2018-"]
```

#### `bar`

```
requires = ["maya>2015,<2019,2017+"]

How should rez resolve this?  There are both 2017 and 2018 "preferred" versions, which are both set to "prefer" in different directions.  I don't think there is a good deterministic way to resolve this unless you start basing it off arbitrary things like package order.

**Comment by mottosso** [comment]
That's true, that would be problematic.

Does the use-case make sense though? What other syntax could work do you think?

**Comment by JeanChristopheMorinPerso** [comment]
I think I agree with @bpabel. While reading I had one thing in mind and it was: How many conflicts would that create in the long run?

Although, I like the idea of having a preferred version. Though not defined in the `package.py` but more in the `rezconfig.py`. BTW, if I'm not mistaken, it already exists (https://github.com/nerdvegas/rez/blob/master/src/rez/package_order.py#L200). Though I never had time to really try it.

There is one particular place where the preferred version could be used and it's the transition to python 3. When you have a lot of packages (and a lot of them are old and didn't anticipated that python3 would really happen in VFX and so they just require `python`, and let's say you don't yet have a `python-3` package yet. Then you add a version 3 and then boom, everything breaks because the packages that were not constraining a version on python would now break. (yep, I'm in this situation and I'm probably not the only one).

**Comment by mottosso** [comment]
Stumbled upon this (from 2013), but haven't dug deeper yet.

https://github.com/nerdvegas/rez/blob/d8c00f6b26eb3289a5ca355398c640e59129fccd/python/rez/rez_config.py#L133-L141

Note: `resolve_mode`

**Comment by bpabel** [comment]
After thinking about this a bit more, it might be simpler to implement if the "preferred" version syntax didn't dictate alternate resolution orders.  In that regards, it's somewhat similar to the "weak" dependencies that are already supported.

```
required = ["maya>=2015,~2018"]   # or whatever the new syntax is
```

The semantics are pretty similar to the original proposal.  If the preferred version doesn't conflict with the normal version dependencies, apply the preferred version in the dependencies as well.  That means you could also do things like this:

```
maya>2016 maya~<2019
```

As long as another dependency doesn't force the preferred dependency to conflict, it will be applied.

It gets much more complex if we want to support multiple preferred version specifiers in the same resolve.  For example

```
maya~>=2017
maya~<2019
maya~2016
```

It's not obvious to me what the correct way to resolve this is.  There's no real way to know if they conflict with the normal dependencies without adding them to the dependencies and doing a resolve.  So, one option would be to resolve each one independently (along with the normal dependencies) and choose whichever preferred version specifier results in the highest ordered version.  

Alternatively, just apply them all in order and raise an exception that they conflict with each other.  

Alternatively, just disallow multiple preferred versions for the same package altogether and raise an exception as soon as multiple preferred versions are found.

**Comment by mottosso** [comment]
Good thoughts.

When it comes to intermixing preferred versions with an overall solve, here's how I'm picturing things at the moment.

1. Preferred versions have no effect throughout the original solve, the solve happening at the moment
2. Once the solve is complete, if and only if two or more versions of a package are possible, e.g. `maya>2015,<2019`, look at what's preferred, e.g. `maya-|2018`.
3. If two or more preferred versions exists, favour one preferred the most number of times
4. If two or more versions are equally preferred, pick the latest or those like usual

I remember seeing talks of specifying a preferred version in the Google Groups some time ago, will see if I can dig that up again, maybe someone somewhere has experience with exactly this and could share some campfire stories.

> Alternatively, just disallow multiple preferred versions for the same package altogether and raise an exception as soon as multiple preferred versions are found.

I wouldn't mind this, to be honest. My particular use-case is specifying preferred versions in a top-level package of sorts, in this case a production project package. In this particular case, there wouldn't be a need for multiple preferred ones.
